File: grail/specs.py
"""The query grammar. docs/grammar.md is the normative BNF; this is it in code.

    query      ::= group ( ws group )*
    group      ::= spec ( ws "^" spec )*
    spec       ::= attr [ "@" range ]
    range      ::= alt ( "||" alt )*
    alt        ::= term ( "," term )*
    term       ::= cmp version | version ".." version
                 | version ".x" | version ".*" | version

A bare version term is a component-wise prefix match, exactly the semantics
mvs solve documents: 3.8 accepts 3.8.9 and refuses 3.81. The upper endpoint
of an interval is prefix-inclusive, so 3.10..3.11 keeps 3.11.9.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from .versions import compare, components

# Comparators, longest first so >= is not read as > followed by =.
_COMPARATORS = (">=", "<=", ">", "<", "=")

# Attrs may be nested-set paths like jetbrains.idea; each segment is plain.
_ATTR_RE = re.compile(r"^[A-Za-z0-9_+-]+(\.[A-Za-z0-9_+-]+)*$")

# Version text is opaque to us beyond the component split.
_VERSION_RE = re.compile(r"^[A-Za-z0-9._+-]+$")


class ParseError(ValueError):
    pass


@dataclass(frozen=True)
class Term:
    """One range term. op is a comparator, 'prefix' or 'interval'."""

    op: str
    version: str
    upper: str | None = None  # interval upper endpoint only

    def matches(self, candidate: str) -> bool:
        if self.op == "prefix":
            return _is_prefix(self.version, candidate)
        if self.op == "interval":
            assert self.upper is not None
            below = compare(candidate, self.upper) <= 0 or _is_prefix(self.upper, candidate)
            return compare(candidate, self.version) >= 0 and below
        c = compare(candidate, self.version)
        return {
            ">=": c >= 0,
            ">": c > 0,
            "<=": c <= 0,
            "<": c < 0,
            "=": c == 0,
        }[self.op]


@dataclass(frozen=True)
class Range:
    """Alternation of conjunctions: any alt may hold, every term in it must."""

    alts: tuple[tuple[Term, ...], ...]

    def matches(self, candidate: str) -> bool:
        return any(all(t.matches(candidate) for t in alt) for alt in self.alts)


@dataclass(frozen=True)
class Spec:
    attr: str
    range: Range | None  # None = any version
    text: str = ""  # the token as typed, for error messages

    def matches(self, candidate: str) -> bool:
        return self.range is None or self.range.matches(candidate)


@dataclass(frozen=True)
class Group:
    """Specs chained with ^: they must resolve at one shared revision."""

    specs: tuple[Spec, ...]


@dataclass(frozen=True)
class Query:
    groups: tuple[Group, ...]

    @property
    def specs(self) -> list[Spec]:
        return [s for g in self.groups for s in g.specs]


def _is_prefix(pattern: str, candidate: str) -> bool:
    """Component-wise prefix: 3.8 accepts 3.8.9 and 3.8, refuses 3.81."""
    p = components(pattern)
    c = components(candidate)
    return len(p) <= len(c) and c[: len(p)] == p


def _parse_term(text: str) -> Term:
    for cmp_op in _COMPARATORS:
        if text.startswith(cmp_op):
            version = text[len(cmp_op) :]
            if not _VERSION_RE.match(version):
                raise ParseError(f"comparator {cmp_op!r} needs a version, got {text!r}")
            return Term(cmp_op, version)

    if ".." in text:
        lo, _, hi = text.partition("..")
        if not (_VERSION_RE.match(lo) and _VERSION_RE.match(hi)):
            raise ParseError(f"interval needs both endpoints, got {text!r}")
        return Term("interval", lo, upper=hi)

    # explicit prefix spellings reduce to the bare-prefix term
    for suffix in (".*", ".x"):
        if text.endswith(suffix):
            text = text[: -len(suffix)]
            break
    if not _VERSION_RE.match(text):
        raise ParseError(f"malformed version term {text!r}")
    return Term("prefix", text)


def _parse_range(text: str) -> Range:
    alts = []
    for alt_text in text.split("||"):
        if not alt_text:
            raise ParseError(f"empty alternative in range {text!r}")
        terms = tuple(_parse_term(t) for t in alt_text.split(","))
        alts.append(terms)
    return Range(tuple(alts))


def _parse_spec(token: str) -> Spec:
    attr, sep, range_text = token.partition("@")
    if not _ATTR_RE.match(attr):
        raise ParseError(f"malformed attribute name {attr!r}")
    if not sep:
        return Spec(attr, None, token)
    if not range_text:
        raise ParseError(f"{attr!r} has @ but no range")
    return Spec(attr, _parse_range(range_text), token)


def parse_query(text: str) -> Query:
    tokens = text.split()
    if not tokens:
        raise ParseError("empty query")

    groups: list[list[Spec]] = []
    for token in tokens:
        chained = token.startswith("^")
        if chained:
            token = token[1:]
            if not token:
                raise ParseError("dangling ^")
            if not groups:
                raise ParseError("^ must follow a spec to chain onto")
        if not token:
            raise ParseError("empty spec")

        spec = _parse_spec(token)
        if chained:
            groups[-1].append(spec)
        else:
            groups.append([spec])

    return Query(tuple(Group(tuple(g)) for g in groups))

File: site/js/specs.js
// The query grammar, twice over: tokenize() classifies every character span
// for the live syntax highlighting, and parseQuery() builds the same AST
// grail/specs.py builds (docs/grammar.md is normative for both).

import { compare, isPrefix } from "./versions.js";

const COMPARATORS = [">=", "<=", ">", "<", "="];
const ATTR_RE = /^[A-Za-z0-9_+-]+(\.[A-Za-z0-9_+-]+)*$/;
const VERSION_RE = /^[A-Za-z0-9._+-]+$/;

export class ParseError extends Error {}

// --- tokenizer, for highlighting -----------------------------------------
// Returns [{text, kind}] covering the input exactly. Kinds: ws, caret,
// attr, at, cmp, version, dots, comma, pipes, error.
export function tokenize(text) {
  const out = [];
  const push = (t, kind) => t && out.push({ text: t, kind });

  for (const piece of text.split(/(\s+)/)) {
    if (!piece) continue;
    if (/^\s+$/.test(piece)) {
      push(piece, "ws");
      continue;
    }

    let rest = piece;
    if (rest.startsWith("^")) {
      push("^", "caret");
      rest = rest.slice(1);
    }
    const at = rest.indexOf("@");
    const attr = at === -1 ? rest : rest.slice(0, at);
    push(attr, ATTR_RE.test(attr) ? "attr" : "error");
    if (at === -1) continue;
    push("@", "at");

    // walk the range, splitting on the operators the grammar knows
    let range = rest.slice(at + 1);
    while (range.length) {
      const op = COMPARATORS.find((c) => range.startsWith(c));
      if (op) {
        push(op, "cmp");
        range = range.slice(op.length);
        continue;
      }
      if (range.startsWith("..")) {
        push("..", "dots");
        range = range.slice(2);
        continue;
      }
      if (range.startsWith("||")) {
        push("||", "pipes");
        range = range.slice(2);
        continue;
      }
      if (range.startsWith(",")) {
        push(",", "comma");
        range = range.slice(1);
        continue;
      }
      const m = range.match(/^[A-Za-z0-9._+*-]+?(?=\.\.|\|\||,|$)/);
      if (!m) {
        push(range, "error");
        break;
      }
      push(m[0], VERSION_RE.test(m[0].replace(/\*$/, "x")) ? "version" : "error");
      range = range.slice(m[0].length);
    }
  }
  return out;
}

// --- parser, mirroring grail/specs.py ------------------------------------

function matchTerm(term, candidate) {
  switch (term.op) {
    case "prefix":
      return isPrefix(term.version, candidate);
    case "interval": {
      const below =
        compare(candidate, term.upper) <= 0 || isPrefix(term.upper, candidate);
      return compare(candidate, term.version) >= 0 && below;
    }
    default: {
      const c = compare(candidate, term.version);
      return { ">=": c >= 0, ">": c > 0, "<=": c <= 0, "<": c < 0, "=": c === 0 }[
        term.op
      ];
    }
  }
}

function parseTerm(text) {
  for (const cmp of COMPARATORS) {
    if (text.startsWith(cmp)) {
      const version = text.slice(cmp.length);
      if (!VERSION_RE.test(version))
        throw new ParseError(`comparator ${cmp} needs a version, got "${text}"`);
      return { op: cmp, version };
    }
  }
  if (text.includes("..")) {
    const [lo, hi] = text.split("..");
    if (!VERSION_RE.test(lo) || !VERSION_RE.test(hi))
      throw new ParseError(`interval needs both endpoints, got "${text}"`);
    return { op: "interval", version: lo, upper: hi };
  }
  let bare = text;
  for (const suffix of [".*", ".x"]) {
    if (bare.endsWith(suffix)) {
      bare = bare.slice(0, -suffix.length);
      break;
    }
  }
  if (!VERSION_RE.test(bare)) throw new ParseError(`malformed version term "${text}"`);
  return { op: "prefix", version: bare };
}

function parseRange(text) {
  const alts = text.split("||").map((alt) => {
    if (!alt) throw new ParseError(`empty alternative in range "${text}"`);
    return alt.split(",").map(parseTerm);
  });
  return {
    alts,
    matches: (candidate) =>
      alts.some((terms) => terms.every((t) => matchTerm(t, candidate))),
  };
}

function parseSpec(token) {
  const at = token.indexOf("@");
  const attr = at === -1 ? token : token.slice(0, at);
  if (!ATTR_RE.test(attr)) throw new ParseError(`malformed attribute name "${attr}"`);
  if (at === -1) return { attr, range: null, text: token };
  const rangeText = token.slice(at + 1);
  if (!rangeText) throw new ParseError(`"${attr}" has @ but no range`);
  return { attr, range: parseRange(rangeText), text: token };
}

export function parseQuery(text) {
  const tokens = text.split(/\s+/).filter(Boolean);
  if (!tokens.length) throw new ParseError("empty query");

  const groups = [];
  for (let token of tokens) {
    const chained = token.startsWith("^");
    if (chained) {
      token = token.slice(1);
      if (!token) throw new ParseError("dangling ^");
      if (!groups.length) throw new ParseError("^ must follow a spec to chain onto");
    }
    const spec = parseSpec(token);
    if (chained) groups[groups.length - 1].push(spec);
    else groups.push([spec]);
  }
  return groups;
}

export function specMatches(spec, version) {
  return spec.range === null || spec.range.matches(version);
}
